feat: call try_merge recursively for list field

mnpw opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #5843.

# Rationale for this change
 
 Merging fields recursively is already performed for a field with datatype Struct. This PR extends that recursive merge to fields with List and LargeList datatypes.
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

# What changes are included in this PR?
- Updated `Field::try_merge` to call merge recursively if datatype is List and LargeList
- Added tests for megre for nested datatypes (like Stuct and List) with Null fields
<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?

N/A
<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

alamb commented on arrow-schema/src/field.rs:
```diff
@@ -483,6 +483,18 @@ impl Field {
                     ));
                 }
             },
+            DataType::List(field) => match &from.data_type {
```
I think we should also add support for `DataType::LargeList` https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#variant.LargeList

If you don't want to add it in this PR, I'll file a ticket to track and add it in a follow on

alamb reviewed (approved):
Thank you for the contribution @mnpw  -- this PR makes sense to me.

Nicely tested as well 👍 

Let me know about `DataType::LargeList`

mnpw commented:
Thanks @alamb for the review. As per your suggestion I have added the change for `LargeList` too.

alamb commented on arrow-schema/src/field.rs:
```diff
@@ -623,6 +645,69 @@ mod test {
+            struct1
+        );
+
+        let mut list1 = Field::new(
```
Any chance we could add a test for `LargeList` too?

alamb reviewed (approved):
Thanks @mnpw 

I noticed that this PR is marked "breaking" but it seems to me like it isn't a breaking API change (in the sense that code that compiles before this PR will also compile after this PR)

mnpw commented:
@alamb 

- Added test for `LargeList`
- Updated commit to reflect the non-breaking nature of change. 

I added `breaking` to highlight the different outcome `Field::try_merge` after my changes. It seems more fit to only add `breaking` if there are user-facing API changes as you suggest.

Lmk if this looks good

alamb commented:
Thanks again @mnpw
